### PR TITLE
Test rewrite in sentry module

### PR DESCRIPTION
**Pull request by metlos**
This is a big PR where each commit corresponds to a rewrite of a single test class from TestNG and JMockit to JUnit and Mockito.

So far, I've been able to avoid constructor or system class mocking which I consider too much magic, because it modifies the bytecode being executed (e.g. you can no longer be sure that what is written in the source code is actually what is executed).

This builds on top of PR https://github.com/getsentry/sentry-java/pull/755 and should not be merged prior to it.

### CODE REVIEW

File: sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java [hunk 1]
```diff
@@ -61,8 +63,10 @@ public AndroidSentryClientFactory(Application app) {
      * Construct an AndroidSentryClientFactory using the specified Android Context.
      *
      * @param ctx Android Context.
+     * @param lookup the lookup for locating the configuration
      */
-    public AndroidSentryClientFactory(Context ctx) {
+    public AndroidSentryClientFactory(Context ctx, Lookup lookup) {
```
Comment (bruno-garcia) [inline sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java]: This would be a breaking change then I guess? We don't offer an overload that doesn't take `Lookup` where we instantiate it ourselves to be backwards compatible?

File: sentry/src/main/java/io/sentry/Sentry.java [hunk 2]
```diff
@@ -1,37 +1,33 @@
...
-     */
-    private static volatile SentryClient storedClient = null;
-    /**
-     * Tracks whether the {@link #init()} method has already been attempted automatically
-     * by {@link #getStoredClient()}.
+     * A synchronization guard of the stored client. Not using the class as the guard because that could theoretically
+     * deadlock with 3rd party code if it also synced on the class.
      */
-    private static AtomicBoolean autoInitAttempted = new AtomicBoolean(false);
+    private static final Object STORED_CLIENT_ACCESS = new Object();
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/Sentry.java]: Is this change needed to get rid of testNG?
Could you please elaborate why move from an atomic operation to a lock?

File: sentry/src/main/java/io/sentry/Sentry.java [hunk 3]
```diff
@@ -67,83 +56,57 @@ public static SentryClient init(SentryClientFactory sentryClientFactory) {
...
-     * @return SentryClient
+     * @deprecated use {@link #init(SentryOptions)}
      */
-    public static SentryClient init(String dsn, SentryClientFactory sentryClientFactory) {
-        SentryOptions sentryOptions = new SentryOptions();
-        sentryOptions.setDsn(dsn);
-        sentryOptions.setSentryClientFactory(sentryClientFactory);
-        return init(sentryOptions);
+    @Deprecated
+    public static SentryClient init(@Nullable String dsn) {
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/Sentry.java]: Is it also needed to deprecate the main init function to get rid of testNG? Maybe this could be a separate PR so we can discuss?

File: sentry/src/main/java/io/sentry/SentryClientFactory.java [hunk 4]
```diff
@@ -1,80 +1,80 @@
...
-     * @param dsn Data Source Name of the Sentry server.
-     * @param sentryClientFactory SentryClientFactory instance to use, or null to do a config lookup.
-     * @return SentryClient instance, or null if one couldn't be constructed.
+     * @param lookup the lookup instance to use for reading the configuration
+     * @param dsn the DSN instance
+     * @return the Sentry client factory or null if not available
      */
-    public static SentryClient sentryClient(String dsn, SentryClientFactory sentryClientFactory) {
-        Dsn realDsn = resolveDsn(dsn);
+    public static @Nullable SentryClientFactory instantiateFrom(Lookup lookup, Dsn dsn) {
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryClientFactory.java]: We wouldn't want to expose `Dsn` type anymore. Generally the SDKs did (and .NET for example still does) but we've taken the approach of exposing the String only which has some advantages like easily binded via configuration files and just simpler for developers/code samples etc.

Also I'd suggest we discuss that on a separate PR unrelated to testNG

File: sentry/src/main/java/io/sentry/SentryOptions.java [hunk 5]
```diff
@@ -1,60 +1,311 @@
...
+     * Creates a new instance of the options using the provided parameters.
+     *
+     * @param lookup        the lookup to locate the configuration with
+     * @param dsn           the dsn to use
+     * @param clientFactory the client factory to use
+     * @throws NullPointerException if any of the parameters is null
+     */
+    public SentryOptions(Lookup lookup, Dsn dsn, SentryClientFactory clientFactory) {
+        this.lookup = requireNonNull(lookup, "lookup");
+        this.dsn = requireNonNull(dsn, "dsn");
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryOptions.java]: An empty dsn is a valid use-case. We just keep the SDK disabled. This is a documented behavior we have in all Sentry SDKs.
Also having a public setter for dsn allows users read their dsn after the options was instantiated. Say a framework that creates the options instance for you and binds values from some configuration source you define.

File: sentry/src/main/java/io/sentry/SentryOptions.java [hunk 6]
```diff
@@ -1,60 +1,311 @@
...
+     * @param lookup  the lookup to locate the configuration with
+     * @param dsn     the optional dsn to use or null if the value should be located in the config
+     * @param factory the client factory to use or null if the value should be located in the config
+     * @return new instance of SentryOptions
+     */
+    public static SentryOptions from(Lookup lookup, @Nullable String dsn, @Nullable SentryClientFactory factory) {
+        return from(lookup, resolveDsn(lookup, dsn), factory);
+    }
+
+    private static SentryOptions from(Lookup lookup, Dsn dsn, @Nullable SentryClientFactory factory) {
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryOptions.java]: By not exposing `Dsn` we get rid of some of these overloads too. Reduce the API surface.

File: sentry/src/main/java/io/sentry/SentryOptions.java [hunk 7]
```diff
@@ -1,60 +1,311 @@
...
+        if (factory == null) {
+            factory = SentryClientFactory.instantiateFrom(lookup, dsn);
+        }
+
+        if (factory == null) {
+            logger.error("Failed to find a Sentry client factory in the provided configuration. Will continue"
+                    + " with a dummy implementation that will send no data.");
+
+            factory = new InvalidSentryClientFactory();
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryOptions.java]: Can we call it `NoOpSentryClientFactory` or something like that? `Invalid` seems like I wouldn't want that instance around.

File: sentry/src/main/java/io/sentry/SentryOptions.java [hunk 8]
```diff
@@ -1,60 +1,311 @@
...
+     * @return the new sentry client or null if the client factory is invalid
      */
-    public ResourceLoader getResourceLoader() {
-        return resourceLoader;
+    @Nullable
+    public SentryClient createClient() {
+        return getClientFactory().createSentryClient(getDsn());
+    }
+
+    private static Dsn resolveDsn(Lookup lookup, @Nullable String dsn) {
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryOptions.java]: Does this belong in `SentryOptions`? In other SDKs we have it more as a POJO type of class.
Perhaps these methods could be pulled out into an internal helper class.

File: sentry/src/main/java/io/sentry/SentryOptions.java [hunk 9]
```diff
@@ -1,60 +1,311 @@
...
+        }
+    }
+
+    // the below is support for instantiating the default lookup
+
+    private static List<ConfigurationResourceLocator> getDefaultResourceLocators() {
+        return asList(new SystemPropertiesBasedLocator(), new EnvironmentBasedLocator(), new StaticFileLocator());
+    }
+
+    private static List<ConfigurationProvider> getDefaultHighPriorityConfigurationProviders() {
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/SentryOptions.java]: The comment above regards POJO applies to these methods too. Would they fit better in some helper?

**Comment by metlos** [comment]
I didn't know how to better create a PR that builds on top of another PR but to just mention it in the description. Most of the changes you discuss in your review are really the changes introduced in PR #755 that these test rewrites build on top.

My intention was to first discuss, finalize and merge  PR #755 and then rebase this thing on top of it. While you were offline for the past couple of days I didn't want to stop my progress and started with the test rewrites present in this draft PR.

**Comment by bruno-garcia** [comment]
I'm sorry I missed #755 altogether. I just assigned myself for review though. Would you like to continue there or address the comments here?

**Comment by metlos** [comment]
> I'm sorry I missed #755 altogether. I just assigned myself for review though. Would you like to continue there or address the comments here?

Let's continue there...

**Comment by metlos** [comment]
This is a huge PR but I took care to do each test class in a single commit, so that should make the reviewing slightly simpler. Also, in this PR only the tests in the `sentry` module are rewritten.
